worlds/mlss/Client.py

```diff
@@ -0,0 +1,292 @@
+            shopping = read_state[9][0] & 0xF
+
+            if logo != "MLSSAP":
+                return
+
+            locs_to_send = set()
+
+            # Checking shop purchases
+            if is_buy:
+                await bizhawk.write(ctx.bizhawk_ctx, [(0x304A, [0x0, 0x0], "EWRAM")])
+                if shop_address != 0x3C0618 and shop_address != 0x3C0684:
+                    location = shop[shop_address][shop_scroll]
+                else:
+                    if shop_init & 0x1 != 0:
+                        location = badge[shop_address][shop_scroll]
+                    else:
+                        location = pants[shop_address][shop_scroll]
+                if location in ctx.server_locations:
+                    locs_to_send.add(location)
+
+            # Loop for receiving items. Item is written as an ID into 0x3057.
+            # ASM reads the ID in a loop and give the player the item before resetting the RAM address to 0x0.
+            # If RAM address isn't 0x0 yet break out and try again later to give the rest of the items
+            for i in range(len(ctx.items_received) - received_index):
+                item_data = items_by_id[ctx.items_received[received_index + i].item]
+                b = await bizhawk.guarded_read(ctx.bizhawk_ctx, [(0x3057, 1, "EWRAM")], [(0x3057, [0x0], "EWRAM")])
+                if b is None:
+                    break
+                await bizhawk.write(
+                    ctx.bizhawk_ctx,
+                    [
+                        (0x3057, [id_to_RAM(item_data.itemID)], "EWRAM"),
+                        (0x4808, [(received_index + i + 1) // 0x100, (received_index + i + 1) % 0x100], "EWRAM"),
+                    ],
+                )
+                await asyncio.sleep(0.1)
+
+            # Early return and location send if you are currently in a shop,
+            # since other flags aren't going to change
+            if shopping & 0x3 == 0x3:
+                if locs_to_send != self.local_checked_locations:
+                    self.local_checked_locations = locs_to_send
+
+                    if locs_to_send is not None:
+                        await ctx.send_msgs([{"cmd": "LocationChecks", "locations": list(locs_to_send)}])
+                return
+
+            # Checking flags that aren't digspots or blocks
+            for item in nonBlock:
+                address, mask, location = item
+                if location in self.local_checked_locations:
+                    continue
+                flag_bytes = await bizhawk.read(ctx.bizhawk_ctx, [(address, 1, "EWRAM"), (0x3060, 6, "EWRAM")])
+                flag_byte = flag_bytes[0][0]
+                backup_logo = bytes([byte for byte in flag_bytes[1] if byte < 0x70]).decode("UTF-8")
+                if backup_logo != "MLSSAP":
+                    return
+                if flag_byte & mask != 0:
+                    if location >= 0xDA0000:
+                        await ctx.send_msgs(
```

Comment:
This is sending a `Set` command for every flag that's set in a list of ~70 flags from what I'm seeing, and it's doing it 8 times per second. Should probably keep a local copy of the bitfield and only update the data storage key when it changes.

And maybe cutting down to only the necessary number of bits if possible, since >53 bits is technically unsupported by JSON.

Reply:
I will be adding local storage variables for the 2 cases later today. In this case as well, only 8 locations have a value >= 0xDA0000.